### ISSUE TITLE
draft timer is not working as expected

### ISSUE THREAD

**Issue by ekievsky** [FlowCrypt/flowcrypt-ios#867]
@tomholub  I have noticed that after moving timer to background it is not working as expected.
better approach needed. I have missed that during initial #514 completion.

**Comment by ekievsky** [comment]
@tomholub I have already started to work on that because it sounds critical for me. I will try to provide fix ASAP.

**Comment by tomholub** [comment]
@ekievsky I've refactored the code completely, so you'd run into merge conflicts. If you want to fix this, please make a branch from my branch here: https://github.com/FlowCrypt/flowcrypt-ios/pull/854 and make a PR to merge into my branch. Unfortunately, it doesn't build yet, so that will make it inconvenient for you. Alternatively do something else first, but we'll make sure to still fix this in 0.4.0.

What exactly is happening as implemented now?

**Comment by ekievsky** [comment]
@tomholub timer is fired only once after latest fix that I pushed ( to make UI not freeze, as far as you remember ). I should test it more properly after thread change, that will be a good lesson.
I am going to find another proper solution and re-test this functionality.

I don't think that there will be much code but if there will - I will make branch from your feature.

**Comment by ekievsky** [comment]
@tomholub I have found out why its working not like we expect.
```
    private func saveDraftIfNeeded() {
        guard shouldSaveDraft() else { return }
        task?.cancel()
        let newTask = Task(priority: .background) {
            do {
                let signingPrv = try await prepareSigningKey()
                let messagevalidationResult = composeMessageService.validateMessage(
                    input: input,
                    contextToSend: contextToSend,
                    email: email,
                    includeAttachments: false,
                    signingPrv: signingPrv
                )
                guard case let .success(message) = messagevalidationResult else {
                    return
                }
                try await composeMessageService.encryptAndSaveDraft(message: message, threadId: input.threadId)
            } catch {}
        }
        task = newTask
    }
```

this is how it should looks like to fire all this functions in background thread.
also, UI methods in `prepareSigningKey` that are using alerts should be converted to separated functions with `@MainActor` modifier which will be the same as using main queue for GCD.

it took a while because I am new to await/async.

making PR to your #854 will not give me a proper chance to test out all this things.
how should we move forward?

**Comment by ekievsky** [comment]
also we should obviously remove GCD code in specific layer if we are using async/await.
`startTimer` can easily work in main thread without blocking UI. main point was of `Task {` instead of `Task(priority: .background) {`

**Comment by tomholub** [comment]
> also we should obviously remove GCD code in specific layer if we are using async/await. `startTimer` can easily work in main thread without blocking UI. main point was of `Task {` instead of `Task(priority: .background) {`

Yes, we're slowly moving in that direction

**Comment by tomholub** [comment]
> making PR to your #854 will not give me a proper chance to test out all this things.
how should we move forward?

You should wait until that PR is merged - work on something else. Else it will just conflict. The code is all different now.

**Comment by ekievsky** [comment]
@tomholub I would like to try to branch from #854 and create a pull request as you proposed above to close this issue.
May I ?

**Comment by tomholub** [comment]
Sure - please try that

**Comment by ekievsky** [comment]
@tomholub I spent some time on your big PR review, it looks like things changes too fast.

right now timer freeze issue is fixed, I believe, by changing threads of underlying functions inside task, but I am not sure. ( too many changes are going on )

I will re-test / re-review on the final version of #854 and close this task, at least to understand what changes I will need to be aware to continue my work.

**Comment by tomholub** [comment]
can retest now

**Comment by ekievsky** [comment]
tested on latest master, working fine for me.

@tomholub would you like to check as well, or I can close this issue?

**Comment by tomholub** [comment]
closing, thanks
